leader-election-impl/src/main/java/com/palantir/paxos/PaxosLearnerImpl.java

```diff
@@ -78,38 +81,35 @@ public PaxosValue getLearnedValue(long seq) {
                     state.put(seq, value);
                 }
             }
-            return state.get(seq);
+            return Optional.ofNullable(state.get(seq));
         } catch (IOException e) {
             logger.error("Unable to get corrupt learned value for sequence {}",
-                    SafeArg.of("sequence", seq), e);
-            return null;
+                    SafeArg.of("sequence", seq),
+                    e);
+            return Optional.empty();
         }
     }
 
     @Override
     public Collection<PaxosValue> getLearnedValuesSince(long seq) {
-        PaxosValue greatestLearnedValue = getGreatestLearnedValue();
-        long greatestSeq = -1L;
-        if (greatestLearnedValue != null) {
-            greatestSeq = greatestLearnedValue.seq;
+        Optional<Long> greatestSeq = getGreatestLearnedValue().map(PaxosValue::getRound);
+        if (!greatestSeq.isPresent()) {
+            return ImmutableList.of();
         }
 
-        Collection<PaxosValue> values = new ArrayList<>();
-        for (long i = seq; i <= greatestSeq; i++) {
-            PaxosValue value;
-            value = getLearnedValue(i);
-            if (value != null) {
-                values.add(value);
-            }
-        }
-        return values;
+        return LongStream.rangeClosed(seq, greatestSeq.get())
+                .boxed()
+                .map(this::getLearnedValue)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(Collectors.toList());
     }
 
     @Override
-    public PaxosValue getGreatestLearnedValue() {
+    public Optional<PaxosValue> getGreatestLearnedValue() {
         if (!state.isEmpty()) {
-            return state.get(state.lastKey());
+            return Optional.ofNullable(state.get(state.lastKey()));
```

Comment:
If I understood correctly that empty result is a "non-standard" path I would invert the if.

Reply:
unrelated to above: this shouldn't be `null` right? since the map is not empty, unless you're thinking it has null values, but non-null keys

Reply:
Agree on inverting if (didn't as I just translated what was there, but yeah it does make sense).

I wrote `Optional.of()` first, but there were some tests that failed. Possibly a bad test, I didn't investigate in detail

Reply:
Sadness is sad